igvm/src/c_api.rs

```diff
@@ -0,0 +1,763 @@
+/// section in a previously parsed IGVM file.
+///
+/// If any error occurs then an [`IgvmApiError`] value is returned instead of
+/// the header type. Errors can be detected by checking if the returned value is
+/// less than zero.
+#[no_mangle]
+pub extern "C" fn igvm_get_header_type(
+    handle: IgvmHandle,
+    section: IgvmHeaderSection,
+    index: u32,
+) -> IgvmResult {
+    // SAFETY: Safe access to the IGVM_HANDLES map through a mutex.
+    let lock = unsafe { IGVM_HANDLES.lock().unwrap() };
+    let igvm_handle = lock.get(&handle);
+
+    // Safety: Using pointer from C
+    if let Some(igvm) = igvm_handle {
+        match section {
+            IgvmHeaderSection::HEADER_SECTION_PLATFORM => {
+                if let Some(header) = igvm.file.platform_headers.get(index as usize) {
+                    header.header_type().0 as IgvmResult
+                } else {
+                    IgvmApiError::IGVMAPI_INVALID_PARAMETER.0
+                }
+            }
+            IgvmHeaderSection::HEADER_SECTION_INITIALIZATION => {
+                if let Some(header) = igvm.file.initialization_headers.get(index as usize) {
+                    header.header_type().0 as IgvmResult
+                } else {
+                    IgvmApiError::IGVMAPI_INVALID_PARAMETER.0
+                }
+            }
+            IgvmHeaderSection::HEADER_SECTION_DIRECTIVE => {
+                if let Some(header) = igvm.file.directive_headers.get(index as usize) {
+                    header.header_type().0 as IgvmResult
+                } else {
+                    IgvmApiError::IGVMAPI_INVALID_PARAMETER.0
+                }
+            }
+            _ => IgvmApiError::IGVMAPI_INVALID_PARAMETER.0,
+        }
+    } else {
+        IgvmApiError::IGVMAPI_INVALID_FILE.0
+    }
+}
+
+/// Prepare a buffer containing the header data in binary form for the entry
+/// with the given index for a particular section in a previously parsed IGVM
+/// file.
+///
+/// The buffer containing the data is returned via a handle from this function.
+/// The handle can be used to access a raw pointer to the data and to query its
+/// size. The buffer handle remains valid until it is closed with a call to
+/// [`igvm_free_buffer()`] or the parsed file handle is closed with a call to
+/// [`igvm_free()`].
+///
+/// If any error occurs then an [`IgvmApiError`] value is returned instead of a
+/// handle. Handles are always greater than zero so a return value greater than
+/// zero indicates a valid handle has been returned.
+#[no_mangle]
+pub extern "C" fn igvm_get_header(
+    handle: IgvmHandle,
+    section: IgvmHeaderSection,
+    index: u32,
+) -> IgvmResult {
+    let mut result = IgvmApiError::IGVMAPI_OK;
+    let mut header_binary = Vec::<u8>::new();
+
+    // SAFETY: Safe access to the IGVM_HANDLES map through a mutex.
+    let mut lock = unsafe { IGVM_HANDLES.lock().unwrap() };
+    let igvm_handle = lock.get_mut(&handle);
+
+    if let Some(igvm) = igvm_handle {
```

Comment:
I'd structure this to help reduce rightward drift in this and other functions by doing something like:
```rust
let igvm = match igvm_handle { // return the valid igvm file or early return from the function with INVALID_HANDLE };
```

That way, we avoid some rightward drift. I think also we can do that for other option usage.

However, looking at this more it seems like maybe we should refactor this so that perhaps we have rust internal functions that use Result<(), IgvmApiError>, then a wrapper conversion function to convert the Result type to the final API call. That way, we can use things like Option `ok_or(..)?` and the like to make this code less C and more Rust.

So we'd have something like:
```rust
pub extern "C" fn igvm_get_header(...) -> IgvmResult {
    match igvm_get_header_internal(...) {
        Ok(()) => IgvmResult::Ok,
        Err(e) => e, 
    }
}
``` 

Then you can write a standard rust function for `igvm_get_header_internal(...) -> Result<(), IgvmResult>`

Reply:
I think we'd find the latter to help reduce a lot of this rightward drift and `if let Some()...` usage throughout.